### PR TITLE
Opening links in external apps

### LINKED ISSUE #136: Open links in corresponding app
## Describe the feature
Links for Slack, Facebook, Instagram, and Email should be opened in the phone's app, if there is one. In case there isn't, a link to a browser should be opened as usual.

## Describe possible solutions considered
The `Linking` component will be used for this and it can be done directly if we have the link to open up the app (it's usually something not straightforward). These links can be found on [deeplink.me](https://www.deeplink.me/) (use the WayBack Machine if broken) or on ~Google~ DuckDuckGo with a bit of research.

```js
// Example for the SHPE Slack workspace
Linking.openURL(slack://open?team=TC61JSPUZ)
```

## Additional comments
This should be simple, but the problem resides in `Linking.canOpenURL`, to check if the user's phone has the app to open it. Without checking if it can open it, the deep link will not do anything (or may even break the app). With Android, it's done pretty straightforward, but iOS is another story, since you need to add some keys into files and do additional changes.

Some stuff that may be helpful...

 **shpeucfapp/ios/shpeucfapp/Info.plist** is the iOS file that contains the following array.

key in "LSApplicationQueriesSchemes" [array](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/LaunchServicesKeys.html#//apple_ref/doc/uid/TP40009250-SW14)

* iOS canOpenUrl](https://developer.apple.com/documentation/uikit/uiapplication/1622952-canopenurl)
* `onPress = { () => Linking.openURL(slack://open?team=TC61JSPUZ) }`
* [Linking component](https://reactnative.dev/docs/linking.html)

### PR DESCRIPTION
**Pull request by grindlock**
Fixes #136 

## Description

Added the functionality to open the linkedin app and mail when button clicked in the profile page. In profile page code to handle app linking for social media buttons added added using uri scheme and deep linking. 

* User page under initial state added linkedin field.
* Forms page under editProfileFormDataRegular added the linkedin field
* Info.plist added the LSApplicationQueriesSchemes array and the whitelisted apps uri to the array.

## Types of changes

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other 

## Checklist:

- [x] My code follows the code style of this project found on [Contribution guidelines](https://github.com/SHPEUCF/shpeucfapp/wiki/Contribution-guidelines).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by grindlock** [comment]
I have been able to test in expo code snip environment and works fine. I was not able to test in the app due to privilege in the database. Is there a test account?

**Comment by grindlock** [comment]
I reopen the pull request after fixing some of the code and updated. Also found a problem that break the app when user click edit profile and then click submit. The problem is Profile under form when passing the value. In form the parameter values is expecting an object but is pass is an array. this is due to formatting  the output to an array before passing it.

File: src/data/FormData.js [hunk 1]
```diff
@@ -161,11 +161,10 @@ const editProfileFormDataRegular = [
 		}
 	},
 	{
-		placeholder: "LinkedIn Profile",
+		placeholder: "LinkedIn Profile (after the in/)",
```
Comment (idelmr) [inline src/data/FormData.js:164]: There should be another way of specifying this, as a non-technical user might not know what "after the in/" means

File: src/screens/User/OtherProfile.js [hunk 2]
```diff
@@ -147,15 +147,20 @@ class OtherProfile extends Component {
...
 			<View style = {{ flex: 0.2 }}>
 				<View style = {{ flex: 0.03 }}></View>
 				<View style = { socialmediarow }>
 					<View style = { [LogoContainer, { backgroundColor: this.props.color, flex: 1 }] }>
 						<TouchableOpacity
 							onPress = { () => {
-								Alert.alert("Coming Soon");
-								// Actions.PostShow({ title: 'Linkedin', uri: 'https://www.linkedin.com/'})
+								this.openApp(linkedin, firstName);
 							} }>
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:164]: Can you move this `onPress` into one line?

File: src/screens/User/OtherProfile.js [hunk 3]
```diff
@@ -164,8 +169,7 @@ class OtherProfile extends Component {
 					<View style = { [LogoContainer, { backgroundColor: this.props.color, flex: 1 }] }>
 						<TouchableOpacity
 							onPress = { () => {
-								Alert.alert("Coming Soon");
-								// Actions.PostShow({ title: 'Github', uri: 'https://www.github.com/'})
+								this.openEmail(email, firstName);
 							} } >
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:173]: One line as well

File: src/screens/User/OtherProfile.js [hunk 4]
```diff
@@ -174,6 +178,42 @@ class OtherProfile extends Component {
 			</View>
 		);
 	}
+
+	async openApp(profile, name) {
+		if (!profile) {
+			Alert.alert(`${ name } has not added his LinkedIn profile yet.`);
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:184]: Change to "...has not added _their_ LinkedIn profile yet"

File: src/screens/User/OtherProfile.js [hunk 5]
```diff
@@ -174,6 +178,42 @@ class OtherProfile extends Component {
...
+		}
+		else {
+			let url = `linkedin://in/${ profile }`;
+
+			console.log(`URL: ${ profile }`);
+			// then have to check if the link can be handle
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:190]: Remove unnecessary comment

File: src/screens/User/OtherProfile.js [hunk 6]
```diff
@@ -174,6 +178,42 @@ class OtherProfile extends Component {
...
 	}
+
+	async openApp(profile, name) {
+		if (!profile) {
+			Alert.alert(`${ name } has not added his LinkedIn profile yet.`);
+		}
+		else {
+			let url = `linkedin://in/${ profile }`;
+
+			console.log(`URL: ${ profile }`);
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:189]: Why is this log here?

File: src/screens/User/OtherProfile.js [hunk 7]
```diff
@@ -174,6 +178,42 @@ class OtherProfile extends Component {
...
+		else {
+			let url = `mailto:${ to }`;
+			const canOpenEmail = await Linking.canOpenURL(url);
+
+			if (canOpenEmail) {
+				await Linking.openURL(url);
+			}
+			else {
+				Alert.alert(`Oops! Somenthing went wrong.\\n
+							If problem persists contact support.`);
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:213]: * Typo, should be "Something"
* Tabs show up on alert

Comment (idelmr) [reply src/screens/User/OtherProfile.js:213]: Actually, change this to:
```js
Alert.alert("Something went wrong, please contact the Tech Director for assistance", { title: "Error", type: "error" });
```

You can change the title to "Oops", "Failure", or anything else that looks best.

File: src/screens/User/Profile.js [hunk 8]
```diff
@@ -29,7 +29,10 @@ class Profile extends Component {
 	render() {
 		return this.renderContent();
 	}
-
+	/*
+	 * using convertObjectToInitialValues() to pass the value back to values break the app due to parameter
+	 * miss match. Form expected an object but receives an array.
+	 */
```
Comment (idelmr) [inline src/screens/User/Profile.js:35]: You changed the values to `this.props.activeUser`. This error that you specify in the comment should be fixed in this PR or create a new issue to solve this, depending on the difficulty. I do not understand the need for a comment

File: src/screens/User/Profile.js [hunk 9]
```diff
@@ -168,23 +171,39 @@ class Profile extends Component {
 		);
 	}
 
-	async openApp(profile){
-		if(profile === "")
-		{
-			Alert.alert("No profile have been added.");
+	async openApp(profile) {
+		if (!profile) {
+			Alert.alert(`No profile have been added.\nTo add your profile click on Edit Profile.`);
```
Comment (idelmr) [inline src/screens/User/Profile.js:176]: Typo, change to "No profile has been added. To add one, tap on Edit Profile"

File: src/screens/User/Profile.js [hunk 10]
```diff
@@ -168,23 +171,39 @@ class Profile extends Component {
 		);
 	}
 
-	async openApp(profile){
-		if(profile === "")
-		{
-			Alert.alert("No profile have been added.");
+	async openApp(profile) {
```
Comment (idelmr) [inline src/screens/User/Profile.js:174]: Specify the app being opened on the function name

File: src/screens/User/Profile.js [hunk 11]
```diff
@@ -168,23 +171,39 @@ class Profile extends Component {
...
+		if (!profile) {
+			Alert.alert(`No profile have been added.\nTo add your profile click on Edit Profile.`);
 		}
-		let url = `linkedin://in/${profile}`;
+		else {
+			let url = `linkedin://in/${profile}`;
 
-		// then have to check if the link can be handle
-		const canOpenLink = await Linking.canOpenURL(url);
+			console.log(`URL: ${profile}`);
```
Comment (idelmr) [inline src/screens/User/Profile.js:181]: Remove log

File: src/screens/User/Profile.js [hunk 12]
```diff
@@ -168,23 +171,39 @@ class Profile extends Component {
...
+			// then have to check if the link can be handle
```
Comment (idelmr) [inline src/screens/User/Profile.js:182]: Remove comment

File: src/screens/User/Profile.js [hunk 13]
```diff
@@ -168,23 +171,39 @@ class Profile extends Component {
...
+				await Linking.openURL(`https://linkedin.com/in/${profile}`);
 		}
-		else
-		{
-			await Linking.openURL(`https://linkedin.com/in/${profile}`);
+	}
+
+	async openEmail(to) {
+		if (!to) {
+			Alert.alert("No email was found.\nTry again again.");
```
Comment (idelmr) [inline src/screens/User/Profile.js:194]: Typo, change to "No email was found. To add one, tap on Edit Profile"

File: src/screens/User/Profile.js [hunk 14]
```diff
@@ -168,23 +171,39 @@ class Profile extends Component {
...
+		else {
+			let url = `mailto:${to}`;
+			const canOpenEmail = await Linking.canOpenURL(url);
+
+			if (canOpenEmail) {
+				await Linking.openURL(url);
+			}
+			else {
+				Alert.alert(`Oops! Somenthing went wrong.\n
+							If problem persists contact support.`);
```
Comment (idelmr) [inline src/screens/User/Profile.js:205]: Use the same alert message from the comment on `OtherProfile.js`

File: src/screens/User/OtherProfile.js [hunk 15]
```diff
@@ -174,6 +178,42 @@ class OtherProfile extends Component {
 			</View>
 		);
 	}
+
+	async openApp(profile, name) {
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:182]: Also specify app being opened

File: src/screens/User/OtherProfile.js (hunk 4 above)
Comment (idelmr) [reply src/screens/User/OtherProfile.js:184]: Remove spaces around curly braces when using interpolation

File: src/screens/User/OtherProfile.js [hunk 16]
```diff
@@ -174,6 +178,42 @@ class OtherProfile extends Component {
...
+		if (!profile) {
+			Alert.alert(`${ name } has not added his LinkedIn profile yet.`);
+		}
+		else {
+			let url = `linkedin://in/${ profile }`;
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:187]: Remove spaces around curly braces

File: src/screens/User/OtherProfile.js [hunk 17]
```diff
@@ -174,6 +178,42 @@ class OtherProfile extends Component {
...
+
+			console.log(`URL: ${ profile }`);
+			// then have to check if the link can be handle
+			const canOpenLink = await Linking.canOpenURL(url);
+
+			if (canOpenLink)
+				await Linking.openURL(url);
+			else
+				await Linking.openURL(`https://linkedin.com/in/${ profile }`);
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:196]: No spaces around curly braces

File: src/screens/User/OtherProfile.js [hunk 18]
```diff
@@ -174,6 +178,42 @@ class OtherProfile extends Component {
...
+		}
+	}
+
+	async openEmail(to, name) {
+		if (!to) {
+			Alert.alert(`No email for ${ name } has been found.`);
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:202]: Consider using the `type: "error"` for alert. Also, remove spacing around curly braces for interpolation

File: src/screens/User/OtherProfile.js [hunk 19]
```diff
@@ -174,6 +178,42 @@ class OtherProfile extends Component {
...
+		}
+		else {
+			let url = `mailto:${ to }`;
```
Comment (idelmr) [inline src/screens/User/OtherProfile.js:205]: Remove spacing around curly brace

**Comment by idelmr** [comment]
Submitting another PR soon

**Pull request closed without merging**